### ISSUE TITLE
React slow to load in production

### ISSUE THREAD

**Issue by mraible** [jhipster/generator-jhipster#7864]
##### **Overview of the issue**

I created an example app, blog post, and screencast about how to develop a Photo Gallery PWA with React and JHipster.

Blog: https://developer.okta.com/blog/2018/06/25/react-spring-boot-photo-gallery-pwa
Screencast: https://youtu.be/GlJWUqy1SJM
Source: https://github.com/oktadeveloper/okta-react-photo-gallery-example
Heroku: https://gallery-pwa.herokuapp.com/

You can see from the screencast and Heroku that i18n translations are slow to load and performance is poor. If I run Lighthouse on Heroku (after it wakes from sleeping), it scores a 39 for performance.

<img width="1396" alt="screen shot 2018-06-26 at 6 28 50 am" src="https://user-images.githubusercontent.com/17892/41911678-8bd823ce-790a-11e8-8185-6f2431309984.png">

##### **Motivation for or Use Case**

I think we can do better.

##### **Reproduce the error**

Create a React app with JHipster 5, deploy it to Heroku, test it with Lighthouse.

##### **JHipster Version(s)**

```
gallery@0.0.0 /Users/mraible/dev/okta/okta-react-photo-gallery-example
└── (empty)

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.okta.developer",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "5.0.1",
    "applicationType": "monolith",
    "baseName": "gallery",
    "packageName": "com.okta.developer",
    "packageFolder": "com/okta/developer",
    "serverPort": "8080",
    "authenticationType": "oauth2",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "clientFramework": "react",
    "useSass": false,
    "clientPackageManager": "yarn",
    "testFrameworks": [
      "protractor"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "fr"
    ],
    "herokuDeployType": "git",
    "herokuAppName": "gallery-pwa"
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Album (album) {
  title String required,
  description TextBlob,
  created Instant
}
entity Photo (photo) {
  title String required,
  description TextBlob,
  image ImageBlob required,
  height Integer,
  width Integer,
  taken Instant,
  uploaded Instant
}
entity Tag (tag) {
  name String required minlength(2)
}
relationship ManyToOne {
  Album{user(login)} to User,
  Photo{album(title)} to Album
}
relationship ManyToMany {
  Photo{tag(name)} to Tag{photo}
}

paginate Album with pagination
paginate Photo, Tag with infinite-scroll

</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

git version 2.14.1

node: v9.8.0

npm: 6.1.0

yeoman: 2.0.2

yarn: 1.7.0

Docker version 18.03.1-ce, build 9ee9f40

docker-compose version 1.21.1, build 5a3f1a3

**Related:** `jhipster info` seems to work for getting the version on Linux, but not on Mac. Notice that it says (empty) for the JHipster version.

**Comment by deepu105** [comment]
Yes definitely,

Dod you use master? can you check if the new lazy loading stuff makes any
difference?

Thanks & Regards,
Deepu

**Comment by mraible** [comment]
I tried with master and it's even slower (22). https://gallery-pwa-master.herokuapp.com/

<img width="1396" alt="screen shot 2018-06-26 at 8 49 15 am" src="https://user-images.githubusercontent.com/17892/41920456-fa19d70c-791d-11e8-9076-f5f0e2203feb.png">

Here's what changed when I upgraded. https://github.com/oktadeveloper/okta-react-photo-gallery-example/compare/jhipster-master

I upgraded by running `rm -rf *` in the directory of my app, then running `jhipster --with-entities`. Then I added back in the customizations that I made for the photo gallery.

**Comment by mraible** [comment]
FWIW, I created my blog demo with JHipster 5, Angular 6, and JWT. It doesn't score very good on performance either. App is at https://jhipster-5-demo.herokuapp.com, performance score seems to be in 40s or 50s.

[Lighthouse Report.pdf](https://github.com/jhipster/generator-jhipster/files/2138345/Lighthouse.Report.pdf)

**Comment by deepu105** [comment]
That's really strange. lazy loading is supposed to give better
performance(btw note that it works only in prod profile)

So the performance is bad only in the lighthouse score or is it bad
visually when using the app as well? I didn't see any performance issue
visually when I was trying them.

I'll try lighthouse when I find time

Thanks & Regards,
Deepu

**Comment by raharrison** [comment]
Loading both React versions isn't great visually for me (clean load). I can clearly see the translation placeholder strings on the landing page whilst the loading bar scrolls and it finally catches up and fills in the correct content.

The Angular version isn't excellent, but still noticeably better. At least I don't see any translation placeholders anywhere. (Note that this might be made worse by my connection - seems to take quite a while to resolve the Heroku apps)

Out of interest does anyone have a report for the old Angular 1 version?

**Comment by atomfrede** [comment]
I have an older jh 4 app here, deployed via dokku on digital ocean (as far as I know). Can only test the login page as I don't have any credentials to that system at hand. It seems to be much better as that app scores 62 for performance (and I'm pretty sure there is much room for improvement in some custom code).

![screenshot-20180626224225-526x761](https://user-images.githubusercontent.com/203401/41938181-a1a828ba-7992-11e8-93b6-323bb74d9a62.png)

**Comment by atomfrede** [comment]
Just a plain bundle analyzer screenshot. I think we should have a look at the font awesome stuff (maybe load it async or so). Also lodash seems pretty large and not sure if we need everything from lodash or lodash at all. Also moment is quite large. Maybe we can check what features we need of moment and maybe replace it with a smaller library.

![screenshot_2018-06-27 webpack bundle analyzer](https://user-images.githubusercontent.com/203401/41954086-afc46868-79d9-11e8-90cc-1ce056ca3097.png)

**Comment by vishal423** [comment]
+1 to upfront load only default translation bundle. I also tried this before, however, noted translation issues on current page. I will see if I can get some time to re-look into that.

Also, we should think about splitting one big bundle to lazy loaded module specific bundles.

**Comment by deepu105** [comment]
Yes we can optimise a lot. For lodash its suppose to treeshake and remove
unused. If its not working then we can just add the methods from lodash
that we are using directly to package instead of full package.

For moment tree shaking doesnt work due to its packaging so ya if there is
a lighter lib, we can consider switching to that.

I'm not happy with the new fontawesome stuff. I would prefer something
thats tree shakeable atleast.

Lets see what we can do.

**Comment by atomfrede** [comment]
Regarding font awesome I will try to update to the latest version later. According to https://github.com/FortAwesome/react-fontawesome/issues/70#issuecomment-399874742 the treeshaking problems should be resolved, which will reduce the bundle size a lo I guess. 

When it comes to lodash there is some 3rd party lib that depends on lodash (maybe it is even jhipster-react). I tried lodash-es and that seems to work fine with regards to treeshaking and imports.

**Comment by mraible** [comment]
For another comparison of v5 vs v4. www.21-points.com is currently running [4.6.2](https://github.com/mraible/21-points/blob/master/.yo-rc.json#L7).

Performance score when not logged in is 23.

<img width="1718" alt="lighthouse-not-logged-in" src="https://user-images.githubusercontent.com/17892/42033424-138941d2-7a9a-11e8-8258-8f24a9d32f78.png">

Performance score when I'm logged in is 17.

<img width="1718" alt="lighthouse-logged-in" src="https://user-images.githubusercontent.com/17892/42033426-15df1e0c-7a9a-11e8-8ad4-e1e2984ab799.png">

**Comment by atomfrede** [comment]
It is very confusing. On my machine the score for the login page is `63`. But maybe a problem of running chromium?

**Comment by atomfrede** [comment]
So maybe I have found the problem why fontawesome is not treeshaked. Let me investigate a little bit further. It seems to be a problem of older babel, uglifyjs and fontawesome. https://github.com/FortAwesome/Font-Awesome/issues/13260

**Comment by deepu105** [comment]
@mraible @atomfrede @vishal423 guys I'm planning to look into this weekend. If you guys have any updates let me know here so that I don't spent time on things you already did/found out

**Comment by atomfrede** [comment]
@deepu105 I will post my findings and what I did yesterday after this evening and maybe propose a PR for the fontawesome treeshaking (which might be easier to see than a comment here).

**Comment by deepu105** [comment]
sure

Thanks & Regards,
Deepu

**Comment by vishal423** [comment]
@deepu105 , I didn't get much chance to look, however, I am sure caching isn't applied on i18n resources. I will move those under content directory. For lazy loading of remaining bundles, I still need to check.

**Comment by deepu105** [comment]
I'll fix lazy loading, i'm already looking into it

Thanks & Regards,
Deepu

**Comment by atomfrede** [comment]
I can't make a proper PR today, so I try to summarize what I did to make font awesome tree shaked. But unfortunately  I just realized the webpack build is not working as it has ts-lint warnings (`no-submodule-imports`). So without adding `@fortawesome` to the whitelist it may be better trying to adapt the webpack config, but here is what I did:

1st update fa deps

```
"@fortawesome/free-solid-svg-icons": "5.1.0",
"@fortawesome/fontawesome-svg-core": "1.2.0",
"@fortawesome/react-fontawesome": "0.1.0",
```
With this all imports must be changed to `import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';` (otherwise no package at all could be created). 

2nd Update icon loader

```
import { faSort } from '@fortawesome/free-solid-svg-icons/faSort';
```

for each icon. --> Fontawesome is correctly tree shaked and the bundle size is decreased a lot.

```
import { faSort } from '@fortawesome/free-solid-svg-icons/faSort';
import { faEye } from '@fortawesome/free-solid-svg-icons/faEye';
import { faSync } from '@fortawesome/free-solid-svg-icons/faSync';
import { faBan } from '@fortawesome/free-solid-svg-icons/faBan';
import { faTrash } from '@fortawesome/free-solid-svg-icons/faTrash';
import { faArrowLeft } from '@fortawesome/free-solid-svg-icons/faArrowLeft';
import { faSave } from '@fortawesome/free-solid-svg-icons/faSave';
import { faPlus } from '@fortawesome/free-solid-svg-icons/faPlus';
import { faPencilAlt } from '@fortawesome/free-solid-svg-icons/faPencilAlt';
import { faUser } from '@fortawesome/free-solid-svg-icons/faUser';
import { faHdd } from '@fortawesome/free-solid-svg-icons/faHdd';
import { faTachometerAlt } from '@fortawesome/free-solid-svg-icons/faTachometerAlt';
import { faHeart } from '@fortawesome/free-solid-svg-icons/faHeart';
import { faList } from '@fortawesome/free-solid-svg-icons/faList';
import { faTasks } from '@fortawesome/free-solid-svg-icons/faTasks';
import { faBook } from '@fortawesome/free-solid-svg-icons/faBook';
import { faClock } from '@fortawesome/free-solid-svg-icons/faClock';
import { faSignInAlt } from '@fortawesome/free-solid-svg-icons/faSignInAlt';
import { faSignOutAlt } from '@fortawesome/free-solid-svg-icons/faSignOutAlt';
import { faThList } from '@fortawesome/free-solid-svg-icons/faThList';
import { faUserPlus } from '@fortawesome/free-solid-svg-icons/faUserPlus';
import { faWrench } from '@fortawesome/free-solid-svg-icons/faWrench';
import { faAsterisk } from '@fortawesome/free-solid-svg-icons/faAsterisk';
import { faFlag } from '@fortawesome/free-solid-svg-icons/faFlag';
import { faBell } from '@fortawesome/free-solid-svg-icons/faBell';
import { faHome } from '@fortawesome/free-solid-svg-icons/faHome';
import { faTimesCircle } from '@fortawesome/free-solid-svg-icons/faTimesCircle';
import { faSearch } from '@fortawesome/free-solid-svg-icons/faSearch';

import { library } from '@fortawesome/fontawesome-svg-core';

export const loadIcons = () => {
  library.add(
    faSort,
    faEye,
    faSync,
    faBan,
    faTrash,
    faArrowLeft,
    faSave,
    faPlus,
    faPencilAlt,
    faUser,
    faTachometerAlt,
    faHeart,
    faList,
    faTasks,
    faBook,
    faHdd,
    faClock,
    faSignInAlt,
    faSignOutAlt,
    faWrench,
    faThList,
    faUserPlus,
    faAsterisk,
    faFlag,
    faBell,
    faHome
  );
};
```

**Comment by vishal423** [comment]
@deepu105, I have some changes for lazy loading of i18n bundles. Let me know if you would like those

**Comment by deepu105** [comment]
Yes please

**Comment by bertwin** [comment]
It may also be worth looking at this problem with uglify:
https://twitter.com/greweb/status/1013122437884542978

**Comment by vishal423** [comment]
@mraible, can you check with latest master as that contain caching improvements along with lazy loading of locale specific bundles.

**Comment by mraible** [comment]
I tried the latest code on master and it still doesn't do well when testing with Lighthouse.

<img width="1399" alt="screen shot 2018-07-02 at 9 16 36 am" src="https://user-images.githubusercontent.com/17892/42172258-bbc583ba-7dd8-11e8-963e-227ed16b007f.png">

**Comment by atomfrede** [comment]
Can you try the tree shaking approach for fontawesome? Just curious how
much that helps

Matt Raible <notifications@github.com> schrieb am Mo., 2. Juli 2018, 17:17:

**Comment by deepu105** [comment]
Guys please check my PR, it should improve things a lot. I'm trying to treeshake lodash now

**Comment by deepu105** [comment]
also RxJS is a huge dependency when enabling websockets.
I dont remember who added it but may be we need to look for a lightweight alternative for that as we dont need all RxJS functionality

**Comment by EmmanuelMess** [comment]
Was this truly fixed, because I am getting <30% pref in lighthouse reports.

**Comment by deepu105** [comment]
@EmmanuelMess the lighthouse reports are heavily dependent on your PC config and network so technically its very difficult to achieve great scores always for everyone. If you identify any improvement that can be done, don't hesitate to PR or tell us

**Comment by mraible** [comment]
I [created an application last week](https://developer.okta.com/blog/2018/10/10/react-native-spring-boot-mobile-app) with JHipster 5.4.2 and React. To test things out, I modified `src/main/webapp/index.html` to enable the service worker, then re-deployed to Cloud Foundry. Attached is my score after I logged in and ran Lighthouse.

<img width="854" alt="screen shot 2018-10-14 at 8 52 22 am" src="https://user-images.githubusercontent.com/17892/46918379-d45d7380-cf8e-11e8-8f1d-dc04fc75c2f3.png">

It seemed strange that I only scored an 81 on PWA since I was able to install the app. So I ran Lighthouse again. The second time it scored 100 for a PWA.

![image](https://user-images.githubusercontent.com/17892/46918399-1a1a3c00-cf8f-11e8-96b2-5d1e85701a43.png)
